### ISSUE TITLE
Bare `Final` special form with dataclasses

### ISSUE THREAD

**Issue by Viicos** [microsoft/pyright#9468]
Using a bare `Final` special form with a default value, pyright will _not_ assume it is a normal field (probably because it assumes it is a `ClassVar`?):

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoBiqAhgDYBQ5okUAJsTMQManEDObApm5trvvUYt2bSgAFBzVh3LCOUAIIAuclDVRiSwiVJQAvFACMq9QCMtRFGQDaqGAF19RyuRCcAbpzIB9eAk4AFAoAdN7eqEgwYQCU5ADEUABEAVykwFoKADRQ5pgo%2BAaG0VAAtAB8UAByYCicieRAA)

```python
from typing import Final

from dataclasses import dataclass

@dataclass
class A:
    a: Final = 1
    b: Final[int] = 1


reveal_type(A.__init__)
# "(self: A, b: int = 1) -> None"
```

However, mypy seems to include `a` as a field ([playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=58b4058c9c7c6a30ddbc8c2d00bc59cf)).

Seems like the [conformance tests](https://github.com/python/typing/blob/main/conformance/tests/dataclasses_final.py) don't cover this use case, so it might have been an oversight.

**Comment by LeeeeT** [comment]
I wonder why python even includes a `Final` field into the `__init__` 🤔

**Comment by Viicos** [comment]
This was clarified in the typing specification [here](https://github.com/python/typing/pull/1669). You can read more about the semantics of the `Final` special form [here](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#semantics-and-examples). While class variables are not included in the `__init__` (because they shouldn't change per instance), `Final` (in the context of a class) means the value can't change _after_ the class have been instantiated.

**Comment by vtgn** [comment]
> I wonder why python even includes a `Final` field into the `__init__` 🤔

In Java for example, you can declare an attribute as Final and define it either directly on the same line, or else in the constructor. It's very useful to be able to set the attribute with a calculated value. It's good Python chose to do the same.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.390.
